### PR TITLE
Extend minikube start command to accept windows os version flag 

### PR DESCRIPTION
**Pull request by bobsira**
Extend the minikube start command to accept the `--windows-node-version` flag for windows node in multi-cluster os.

[Windows Node Support in minikube Proposal](https://github.com/kubernetes/minikube/pull/18707)

### CODE REVIEW

**Comment by iankingori** [comment]
Update docs to highlight this change as well

**Review by iankingori — APPROVED** [review]
LGTM

File: pkg/minikube/node/start.go [hunk 1]
```diff
@@ -972,3 +972,8 @@ To see benchmarks checkout https://minikube.sigs.k8s.io/docs/benchmarks/cpuusage
 `, out.V{"drivers": altDriverList.String()})
 	}
 }
+
+// ValidWindowsOS lists the supported Windows OS versions
```
Comment (spowelljr) [inline pkg/minikube/node/start.go]:
```suggestion
// ValidWindowsOSVersions lists the supported Windows OS versions
```

File: cmd/minikube/cmd/start_flags.go [hunk 2]
```diff
@@ -208,6 +209,8 @@ func initMinikubeFlags() {
 	startCmd.Flags().String(staticIP, "", "Set a static IP for the minikube cluster, the IP must be: private, IPv4, and the last octet must be between 2 and 254, for example 192.168.200.200 (Docker and Podman drivers only)")
 	startCmd.Flags().StringP(gpus, "g", "", "Allow pods to use your NVIDIA GPUs. Options include: [all,nvidia] (Docker driver with Docker container-runtime only)")
 	startCmd.Flags().Duration(autoPauseInterval, time.Minute*1, "Duration of inactivity before the minikube VM is paused (default 1m0s)")
+	startCmd.Flags().String(windowsNodeVersion, constants.DefaultWindowsNodeVersion, "The version of Windows to use for the windows node on a multi-node cluster (e.g., 2019, 2022). Defaults to Windows Server 2022")
```
Comment (spowelljr) [inline cmd/minikube/cmd/start_flags.go]:
```suggestion
	startCmd.Flags().String(windowsNodeVersion, constants.DefaultWindowsNodeVersion, "The version of Windows to use for the Windows node on a multi-node cluster (e.g., 2019, 2022). Defaults to Windows Server 2022")
```

File: cmd/minikube/cmd/start.go [hunk 3]
```diff
@@ -1303,6 +1303,13 @@ func validateFlags(cmd *cobra.Command, drvName string) { //nolint:gocyclo
 		validateCNI(cmd, viper.GetString(containerRuntime))
 	}
 
+	if cmd.Flags().Changed(windowsNodeVersion) {
+		err := validateWindowsOSVersion(viper.GetString(windowsNodeVersion))
+		if err != nil {
```
Comment (spowelljr) [inline cmd/minikube/cmd/start.go]:
```suggestion
		if err := validateWindowsOSVersion(viper.GetString(windowsNodeVersion)); err != nil {
```

File: pkg/minikube/node/start.go [hunk 4]
```diff
@@ -972,3 +972,8 @@ To see benchmarks checkout https://minikube.sigs.k8s.io/docs/benchmarks/cpuusage
...
+func ValidWindowsOSVersions() []string {
+	return []string{"2019", "2022"}
```
Comment (spowelljr) [inline pkg/minikube/node/start.go]:
```suggestion
func ValidWindowsOSVersions() map[string]bool {
	return map[string]bool{"2019": true, "2022": true}
```

Would be easier to check this as a map

File: cmd/minikube/cmd/start.go [hunk 5]
```diff
@@ -1424,6 +1431,28 @@ func validateDiskSize(diskSize string) error {
...
+		if osVersion == option {
+			validOSVersion = true
+		}
+	}
+
+	if !validOSVersion {
+		return errors.Errorf("Invalid Windows Server OS Version: %s. Valid OS version are: %s", osVersion, node.ValidWindowsOSVersions())
+	}
+
+	return nil
```
Comment (spowelljr) [inline cmd/minikube/cmd/start.go]:
```suggestion
	if validOptions[osVersion] {
		return nil
	}

	return errors.Errorf("Invalid Windows Server OS Version: %s. Valid OS version are: %s", osVersion, maps.Keys(validOptions))
```

Using the map mentioned above greatly simplifies this

**Review by spowelljr — CHANGES_REQUESTED** [review]

File: cmd/minikube/cmd/start_test.go [hunk 6]
```diff
@@ -477,6 +477,38 @@ func TestValidateRuntime(t *testing.T) {
...
+	}
+	for _, test := range tests {
+		t.Run(test.osVersion, func(t *testing.T) {
+			got := validateWindowsOSVersion(test.osVersion)
+			gotError := ""
+			if got != nil {
+				gotError = got.Error()
+			}
+			if gotError != test.errorMsg {
+				t.Errorf("ValidateWindowsOSVersion(osVersion=%v): got %v, expected %v", test.osVersion, got, test.errorMsg)
```
Comment (spowelljr) [inline cmd/minikube/cmd/start_test.go:506]:
```suggestion
				t.Errorf("ValidateWindowsOSVersion(osVersion=%s): got %v, expected %v", test.osVersion, got, test.errorMsg)
```

Comment (spowelljr) [inline cmd/minikube/cmd/start_test.go:506]:
```suggestion
		wantedError  error
	}{
		{
			osVersion: "2019",
		},
		{
			osVersion: "2022",
		},
		{
			osVersion: "2023",
			wantedError:  errors.New("Invalid Windows Server OS Version: 2023. Valid OS version are: [2019 2022]"),
		},
	}
	for _, tc := range tests {
		t.Run(tc.osVersion, func(t *testing.T) {
			got := validateWindowsOSVersion(tc.osVersion)
			if got != tc.wantedError {
				t.Errorf("ValidateWindowsOSVersion(osVersion=%s): got %v, expected %v", test.osVersion, got, tc.wantedError)
```

**Comment by spowelljr** [comment]
> Update docs to highlight this change as well

Not sure if you're aware, but you can update the docs automatically via `make generate-docs`

**Comment by bobsira** [comment]
> > Update docs to highlight this change as well
> 
> Not sure if you're aware, but you can update the docs automatically via `make generate-docs`

Thanks for pointing this out! I was not aware of this!

File: pkg/minikube/node/start.go (hunk 4 above)
Comment (bobsira) [reply pkg/minikube/node/start.go]: for some reason the unit test is failing with this as a map. I am not sure the return map will always have 2019 first then 2022. === test failure message -> Invalid Windows Server OS Version: 2023. Valid OS version are: [2019 2022], expected Invalid Windows Server OS Version: 2023. Valid OS version are: [2019 2022]

```
func maps.Keys(m map[string]bool) []string // func[M ~map[K]V, K comparable, V any](m M) []K
Keys returns the keys of the map m. The keys will be in an indeterminate order.
```

Comment (spowelljr) [reply pkg/minikube/node/start.go]: You just need to add the `maps` to the import at the top of the file.

Comment (spowelljr) [reply pkg/minikube/node/start.go]: I see, yeah that will be a problem. The easiest would likely be to change it back to a `[]string`, but you can simplify the loop you had before, if `option` equals the version you're looking for you can just return `nil`. Another option is you could sort the slice before outputting it with `slices.Sort` which would also resolve your issue, but the former is likely a better option. The overkill option if you make a `type` (`WindowsOSVersions`) and you add a `String()` to it which does the `maps.Keys` and `slices.Sort` but is probably a bit overkill for what you need.

File: cmd/minikube/cmd/start_test.go [hunk 7]
```diff
@@ -477,6 +477,38 @@ func TestValidateRuntime(t *testing.T) {
...
+			osVersion: "2019",
+			errorMsg:  "",
+		},
+		{
+			osVersion: "2022",
+			errorMsg:  "",
+		},
+		{
+			osVersion: "2023",
+			errorMsg:  "Invalid Windows Server OS Version: 2023. Valid OS version are: [2019 2022]",
```
Comment (spowelljr) [inline cmd/minikube/cmd/start_test.go:495]: The failure comparison will be much simpler with `errors.New`, then you can compare directly check if `got != test.wantedError`

**Review by spowelljr — CHANGES_REQUESTED** [review]

File: cmd/minikube/cmd/start_test.go (hunk 7 above)
Comment (bobsira) [reply cmd/minikube/cmd/start_test.go:495]: @spowelljr  I did realize that `errors.New` creates a new error value each time it's called so the comparison fails and 

How does this alternative implementation look? 

```
func TestValidateWindowsOSVersion(t *testing.T) {
	var tests = []struct {
		osVersion   string
		wantedError error
	}{
		{
			osVersion:   "2019",
			wantedError: nil,
		},
		{
			osVersion:   "2022",
			wantedError: nil,
		},
		{
			osVersion:   "2023",
			wantedError: errors.New("Invalid Windows Server OS Version: 2023. Valid OS version are: [2019 2022]"),
		},
	}
	for _, test := range tests {
		t.Run(test.osVersion, func(t *testing.T) {
			got := validateWindowsOSVersion(test.osVersion)
			if got != nil && test.wantedError != nil {
				if got.Error() != test.wantedError.Error() {
					t.Errorf("ValidateWindowsOSVersion(osVersion=%s): got %v, expected %v", test.osVersion, got, test.wantedError)
				}
			} else if got != test.wantedError {
				t.Errorf("ValidateWindowsOSVersion(osVersion=%s): got %v, expected %v", test.osVersion, got, test.wantedError)
			}
		})
	}
}
```

The first `if` block checks if both `got `and `wantedError `are not `nill`. If true, it then compares the error messgaes of `got `and `wantedError `using their `Error()` methods. 
The `else if` block handles cases where the expectations about the error presence differ: either `got `or `wantedError `is nil, but not both, or both are `nil`.

This  passes the test! Let me know if this looks alright to you and I'll push it upstream.

**Review by TinaMor — APPROVED** [review]

File: cmd/minikube/cmd/start_test.go (hunk 7 above)
Comment (spowelljr) [reply cmd/minikube/cmd/start_test.go:495]: Sorry, I'm used to us wrapping errors, I see that just creating a new error doesn't work. What you have is good then

**Review by spowelljr — APPROVED** [review]

**Pull request merged**